### PR TITLE
apogee submission (0.2293)

### PR DESCRIPTION
**Pull request by adpena**
# submission name:
apogee

# upload zipped `archive.zip`
https://github.com/adpena/comma_video_compression_challenge/releases/download/apogee-pr98-hnerv-adapter-20260504/archive.zip

# report.txt
```
=== Evaluation config ===
  batch_size: 16
  device: cuda
  num_threads: 2
  prefetch_queue_depth: 4
  seed: 1234
  submission_dir: submissions/apogee
=== Evaluation results over 600 samples ===
  Average PoseNet Distortion: 0.00017394
  Average SegNet Distortion: 0.00068841
  Submission file size: 178,392 bytes
  Original uncompressed size: 37,545,489 bytes
  Compression Rate: 0.00475136
  Final score: 100*segnet_dist + sqrt(10*posenet_dist) + 25*rate = 0.23

Exact local CUDA/T4 custody:
  score_recomputed_from_components: 0.22933111465960354
  archive_sha256: 7ecb0df1c4627d55d88e03eff3d890b7a7a5b047c62515acff20232cf29310eb
  archive_size_bytes: 178392
  archive_member: 0.bin
  archive_member_sha256: fce200db2fe087cc6a051945b3fda2c37f5bbb3e19b8f20a1aea7201db0c9f5f
  n_samples: 600
  eval_hardware: Tesla T4
  runtime_tree_sha256: 0232154c17410621325ec1647e0f0723b3310d63b0d4bc4bf7bbb5e9aa2fccd0
```

# does your submission require gpu for evaluation (inflation)?
yes

# did you include the compression script? and want it to be merged?
no

# additional comments
Apogee is a contest-faithful HNeRV submission with the exact archive bytes
hosted as a release asset and the deterministic inflate runtime included in
this PR.

Public artifacts:
- PR branch: https://github.com/adpena/comma_video_compression_challenge/tree/apogee-pr98-hnerv-adapter/submissions/apogee
- Release asset: https://github.com/adpena/comma_video_compression_challenge/releases/tag/apogee-pr98-hnerv-adapter-20260504
- Runtime entrypoint: `submissions/apogee/inflate.sh`
- Archive manifest: `submissions/apogee/archive_manifest.json`

The submission was validated locally through the canonical path:

```
archive.zip -> inflate.sh -> upstream/evaluate.py
```

We also ran a strict pre-submission compliance gate before opening this PR.
That gate checked archive SHA/size, single-member ZIP integrity,
local-header/central-directory consistency, runtime tree custody, report
linkage, public hygiene, T4-equivalent CUDA auth eval, and exact
component-score recomputation.

Design summary:
- HNeRV-style learned decoder and per-pair latent stream optimized for the
  fixed contest video and task scorers.
- Deterministic runtime wrapper with a closed one-video contract for `0.mkv`.
- All score-affecting bytes are charged inside `archive.zip`; runtime source
  is only the decoder/inflate program.
- Score claims are tied to exact archive SHA-256 and runtime-tree SHA-256 rather than human logs.

Methodology and meta-analysis:
- The challenge objective was treated as a rate-distortion compiler problem:
  representation, decoder, quantization, side information, and packing are
  optimized against the exact SegNet/PoseNet/rate formula.
- We tracked both positive and negative evidence. Failed archive variants,
  wrapper bugs, CPU/MPS/proxy disagreements, and byte-only wins that harmed
  PoseNet/SegNet were used as guardrails rather than score claims.
- Public frontier submissions showed that the winning basin is not a generic
  perceptual video codec; it is a compact machine-score representation for one
  fixed video and two fixed neural evaluators. Apogee follows that principle
  while keeping all score-affecting state inside the charged archive.
- The archive is intentionally simple at the outer layer: one ZIP member,
  deterministic byte identity, no hidden sidecars, no scorer patches, and a
  runtime that fails closed on the expected one-video contest contract.

Public source for this submitted runtime is in this PR. The archive asset is immutable at the release URL above.

## Production and OSS deployment notes

This PR intentionally keeps the contest runtime small and auditable rather than
shipping the full research workspace. The production-relevant properties are:

- deterministic single-payload archive: one `0.bin` member, fixed SHA-256, no
  hidden sidecars or local paths;
- explicit runtime custody: the submitted runtime tree is small enough to audit
  file-by-file and was hashed during exact evaluation;
- deployable inference shape: `inflate.sh` is the only entrypoint and delegates
  to a compact Python/PyTorch decoder;
- hardware fit: local exact CUDA/T4 auth eval completed in about one minute,
  with inflate itself taking about 22.6 seconds, well inside the 30-minute
  contest limit;
- reproducible validation: score claims are tied to exact archive SHA, runtime
  tree SHA, component distances, sample count, and T4 hardware evidence;
- clean separation of concerns: the research system (`tac`) produced and
  validated the archive, while this PR exposes only the minimal runtime needed
  by comma's evaluator.

A productionized version inside comma/openpilot should preserve this boundary:
keep the trained representation and entropy-coded payload as charged data,
keep the inflate path deterministic and side-effect-light, and move reusable
codec components into a small package with typed payload contracts, CI replay
fixtures, and byte-for-byte archive tests. The broader research tooling can be
released separately after sanitizing provider state, local custody ledgers, and
operator-specific paths.

---

## OSS extraction status (updated 2026-05-05)

Two repos hold the work behind this submission:

**`tac`** — *Task-Aware Compression* — the tight production library:
https://github.com/adpena/tac (MIT, 214 source files, 415 tests, CI-passing
on Python 3.11/3.12). This is the reusable codec primitives, predictor with
refusal modes, distortion proxy, parallel-dispatch toolchain, and hardened
preflight infrastructure used to produce and validate this submission.
Importable via `pip install -e .` from a clone.

**`comma-lab`** — *the comma video compression contest research workspace* —
the comprehensive research record:
https://github.com/adpena/comma-lab (private pending sanitization). This is
the parent directory containing the full experimental ledger, dispatch
wrappers, lane registry, byte-level deconstruction of all 54 scored public
PRs, methodology writeup (`docs/paper/01_introduction.md` through
`docs/paper/07_discussion.md`), and ~585 session-memory files capturing the
research trajectory, findings, failures, and bug-class extinctions. The
sanitization pass strips ~30GB of regenerable cache files (uncompressed
`0.raw` ground-truth frame dumps, `gt_frames.pt`/`comp_frames.pt` training
caches, `renderer_output_for_postfilter.pt`, decoded mask tensors, the
`raft_flow.pt` optical flow cache) — all re-derivable from the source
archive, none containing unique research signal. Public release expected
within days.

**Honest accounting:** this OSS extraction work was deprioritized during the
all-night sub-deadline scramble for leaderboard position, and is now being
done post-deadline. Future PRs to this challenge (if comma runs adjacent
contests) will demonstrate the closed-loop parallel-dispatch toolchain that
was missing during this contest's race window — see `docs/paper/07_discussion.md`
§7.8 in `comma-lab` for the postmortem of how the May 4 race window was
decided in 4h 8min after PR #95 (HNeRV root) published, with the silver
medal landing in 241 lines / 2 files / under 4 hours.

### CODE REVIEW

**Comment by YassineYousfi** [comment]
@adpena Congratulations on making it onto the leaderboard! I hope you had fun competing in the challenge and learned something new along the way. If you are looking for a job or internship, please email givemeajob@comma.ai with a link to this PR. See you in the next challenge!
